packages/signers/README.md

```diff
@@ -0,0 +1,368 @@
+
+### Types
+
+#### `KeyPairSigner<TAddress>`
+
+Defines a signer that uses a `CryptoKeyPair` to sign messages and transactions. It implements both the `MessagePartialSigner` and `TransactionPartialSigner` interfaces and keeps track of the `CryptoKeyPair` instance used to sign messages and transactions.
+
+```ts
+import { generateKeyPairSigner } from '@solana/signers';
+
+const myKeyPairSigner = generateKeyPairSigner();
+myKeyPairSigner.address; // Address;
+myKeyPairSigner.keyPair; // CryptoKeyPair;
+const [myMessageSignatures] = await myKeyPairSigner.signMessage([myMessage]);
+const [myTransactionSignatures] = await myKeyPairSigner.signTransaction([myTransaction]);
+```
+
+### Functions
+
+#### `createSignerFromKeyPair()`
+
+Creates a `KeyPairSigner` from a provided Crypto KeyPair. The `signMessage` and `signTransaction` functions of the returned signer will use the private key of the provided key pair to sign messages and transactions. Note that both the `signMessage` and `signTransaction` implementations are parallelized, meaning that they will sign all provided messages and transactions in parallel.
+
+```ts
+import { generateKeyPair } from '@solana/keys';
+import { createSignerFromKeyPair, KeyPairSigner } from '@solana/signers';
+
+const myKeyPair: CryptoKeyPair = await generateKeyPair();
+const myKeyPairSigner: KeyPairSigner = await createSignerFromKeyPair(myKeyPair);
+```
+
+#### `generateKeyPairSigner()`
+
+A convenience function that generates a new Crypto KeyPair and immediately creates a `KeyPairSigner` from it.
+
+```ts
+import { generateKeyPairSigner } from '@solana/signers';
+
+const myKeyPairSigner = await generateKeyPairSigner();
+```
+
+#### `isKeyPairSigner()`
+
+A type guard that returns `true` if the provided value is a `KeyPairSigner`.
+
+```ts
+const myKeyPairSigner = await generateKeyPairSigner();
+isKeyPairSigner(myKeyPairSigner); // ✅ true
+isKeyPairSigner({ address: address('1234..5678') }); // ❌ false
+```
+
+#### `assertIsKeyPairSigner()`
+
+A type guard that throws an error if the provided value is not a `KeyPairSigner`.
+
+```ts
+const myKeyPairSigner = await generateKeyPairSigner();
+assertIsKeyPairSigner(myKeyPairSigner); // ✅ void
+assertIsKeyPairSigner({ address: address('1234..5678') }); // ❌ Throws an error.
+```
+
+## Creating Noop signers
+
+### Functions
+
+#### `createNoopSigner()`
+
+_Coming soon..._
+
+Creates a Noop (No-Operation) signer from a given address. It will return an implementation of both the `MessagePartialSigner` and `TransactionPartialSigner` interfaces that do not sign anything. Namely, signing a transaction or a message will return an empty `SignatureDictionary`.
+
+This signer may be useful:
+
+-   For testing purposes.
+-   For indicating that a given account is a signer and taking the responsibility to provide the signature for that account ourselves. For instance, if we need to send the transaction to a server that will sign it and send it for us.
```

Comment:
I know you haven't included this in the code, but I wanted to mention I don't really love the naming convention "no-op" here.

This sounds like the type of signer that would be used primarily with transaction requests, and I think it should have a name that's less off-putting for folks using Solana Pay, etc. and marking signers whose signatures might be provided by a QR-scanning wallet (for example).

My mind goes to "IOUSigner" (lol, you get the point), but how about something like `FutureSigner`?

Reply:
Hmm I agree that `NoopSigner` feels low-level and not as inviting as it could be. However it best describes what the signer does: nothing.

I worry that with something like `FutureSigner` people will assume the signer actually does something.

Reply:
Yes but it's doing nothing and is expecting resolution at some point. It's sort of like ear-marking a message/transaction for a later signature.

Reply:
Let's move this conversation [here](https://github.com/solana-labs/solana-web3.js/pull/1842).